### PR TITLE
YAML shuffle tests for v0.6

### PR DESCRIPTION
**Pull request by shayzluf**
[Part of] #2307 

---

# YAML Shuffle tests

in order to test that we comply with the spec running the latest spec YAML tests is a must

**Write a summary of the changes you are making**
changes in:
beacon-chain/utils/shuffle.go
shared/bytesutil/bytes.go
shared/bytesutil/bytes_test.go

**Link anything that would be helpful or relevant to the reviewers**
https://github.com/ethereum/eth2.0-spec-tests

### CODE REVIEW

File: beacon-chain/chaintest/backend/simulated_backend.go [hunk 1]
```diff
@@ -177,17 +177,35 @@ func (sb *SimulatedBackend) RunForkChoiceTest(testCase *ForkChoiceTestCase) erro
...
+		}
+		shuffledList[si] = i
 	}
-	if !reflect.DeepEqual(output, testCase.Output) {
-		return fmt.Errorf("shuffle result error: expected %v, actual %v", testCase.Output, output)
+	log.Infof("shuffledIndexList: %v", shuffledList)
+	if !reflect.DeepEqual(shuffledList, testCase.Shuffled) {
+		return fmt.Errorf("shuffle result error: expected %v, actual %v", testCase.Shuffled, shuffledList)
 	}
+	// sl, err := utils.ShuffleList(testIndices, seed)
```
Comment (nisdas) [inline beacon-chain/chaintest/backend/simulated_backend.go]: why is this commented out ?

File: beacon-chain/chaintest/main.go [hunk 2]
```diff
@@ -77,19 +88,28 @@ func runTests(tests []interface{}, sb *backend.SimulatedBackend) error {
...
 			log.Infof("Title: %v", typedTest.Title)
 			log.Infof("Summary: %v", typedTest.Summary)
-			log.Infof("Test Suite: %v", typedTest.TestSuite)
-			log.Infof("Fork: %v", typedTest.Fork)
-			log.Infof("Version: %v", typedTest.Version)
+			log.Infof("Fork: %v", typedTest.Forks)
+			log.Infof("Config: %v", typedTest.Config)
+			config, ok := configs[string(typedTest.Config)]
+			if !ok {
+				return fmt.Errorf("no config file found for test")
```
Comment (nisdas) [inline beacon-chain/chaintest/main.go]: no need for formatting if there is no variable to log

File: beacon-chain/chaintest/main.go [hunk 3]
```diff
@@ -77,19 +88,28 @@ func runTests(tests []interface{}, sb *backend.SimulatedBackend) error {
...
+			}
+			conf, ok := config.(*params.BeaconChainConfig)
+
+			if !ok {
+				return fmt.Errorf("config file is not of type *params.BeaconChainConfig")
```
Comment (nisdas) [inline beacon-chain/chaintest/main.go]: log the config file type here

File: beacon-chain/utils/shuffle_test.go [hunk 4]
```diff
@@ -74,19 +66,105 @@ func TestSplitIndices_OK(t *testing.T) {
...
+	for i := uint64(0); i < listSize; i++ {
+		si, err := ShuffledIndex(i, listSize, seed)
+		if err != nil {
+			t.Error(err)
+		}
+		shuffledList[si] = i
+	}
+	t.Logf("shuffledList: %v", shuffledList)
+	for i := uint64(0); i < listSize; i++ {
+		ui, err := UnShuffledIndex(i, listSize, seed)
```
Comment (nisdas) [inline beacon-chain/utils/shuffle_test.go]: This test doesnt look right, why are you shuffling and unshuffling on the same index ? shouldnt you unshuffle the shuffled list ?

File: beacon-chain/chaintest/backend/simulated_backend.go (hunk 1 above)
Comment (shayzluf) [reply beacon-chain/chaintest/backend/simulated_backend.go]: removed commented code
thanks

File: beacon-chain/chaintest/main.go (hunk 2 above)
Comment (shayzluf) [reply beacon-chain/chaintest/main.go]: fixed

File: beacon-chain/chaintest/main.go (hunk 3 above)
Comment (shayzluf) [reply beacon-chain/chaintest/main.go]: logged thanks

File: beacon-chain/utils/shuffle_test.go (hunk 4 above)
Comment (shayzluf) [reply beacon-chain/utils/shuffle_test.go]: not readable but it was doing the same. changed it

Comment (shayzluf) [reply beacon-chain/utils/shuffle_test.go]: checking again

Comment (shayzluf) [reply beacon-chain/utils/shuffle_test.go]: i think it is ok it re arrange the shuffled array by the un-shuffle index

**Comment by shayzluf** [comment]
spec tests still fail. corresponding about it on gitter (hoping to get it solved soon)

File: beacon-chain/chaintest/backend/simulated_backend.go [hunk 5]
```diff
@@ -9,12 +9,13 @@ import (
 	"reflect"
 	"time"
 
+	"github.com/prysmaticlabs/prysm/beacon-chain/utils"
+
```
Comment (terencechain) [inline beacon-chain/chaintest/backend/simulated_backend.go]: space

File: beacon-chain/utils/shuffle.go [hunk 6]
```diff
@@ -80,62 +33,77 @@ func SplitIndices(l []uint64, n uint64) [][]uint64 {
...
 }
 
-// PermutedIndex returns `p(index)` in a pseudorandom permutation `p` of `0...list_size - 1` with ``seed`` as entropy.
+// ShuffledIndex returns `p(index)` in a pseudorandom permutation `p` of `0...list_size - 1` with ``seed`` as entropy.
 // We utilize 'swap or not' shuffling in this implementation; we are allocating the memory with the seed that stays
 // constant between iterations instead of reallocating it each iteration as in the spec. This implementation is based
 // on the original implementation from protolambda, https://github.com/protolambda/eth2-shuffle
-//
+func ShuffledIndex(index uint64, indexCount uint64, seed [32]byte) (uint64, error) {
+	return innerShuffledIndex(index, indexCount, seed, true)
```
Comment (terencechain) [inline beacon-chain/utils/shuffle.go]:
```suggestion
	return innerShuffledIndex(index, indexCount, seed, true /* dir */)
```

I suggest a better name than `dir` not sure what it means in this context

File: beacon-chain/utils/shuffle.go [hunk 7]
```diff
@@ -80,62 +33,77 @@ func SplitIndices(l []uint64, n uint64) [][]uint64 {
...
-	if listSize > maxShuffleListSize {
+	if indexCount > maxShuffleListSize {
 		return 0, fmt.Errorf("list size %d out of bounds",
-			listSize)
+			indexCount)
+	}
+	rounds := uint8(params.BeaconConfig().ShuffleRoundCount)
+	round := uint8(0)
+	if !dir {
+		// Start at last round.
```
Comment (terencechain) [inline beacon-chain/utils/shuffle.go]:
```suggestion
		// Starting last round and iterating through the rounds in reverse, un-swaps everything, effectively un-shuffling the list.
```

File: beacon-chain/utils/shuffle.go [hunk 8]
```diff
@@ -80,62 +33,77 @@ func SplitIndices(l []uint64, n uint64) [][]uint64 {
...
+		// Iterating through the rounds in reverse, un-swaps everything, effectively un-shuffling the list.
```
Comment (terencechain) [inline beacon-chain/utils/shuffle.go]:
```suggestion
```

File: beacon-chain/utils/shuffle.go [hunk 9]
```diff
@@ -161,3 +142,140 @@ func PermutedIndex(index uint64, listSize uint64, seed [32]byte) (uint64, error)
...
+
+// ShuffleList returns list of shuffled indexes in a pseudorandom permutation `p` of `0...list_size - 1` with ``seed`` as entropy.
+// We utilize 'swap or not' shuffling in this implementation; we are allocating the memory with the seed that stays
+// constant between iterations instead of reallocating it each iteration as in the spec. This implementation is based
+// on the original implementation from protolambda, https://github.com/protolambda/eth2-shuffle
+func ShuffleList(input []uint64, seed [32]byte) ([]uint64, error) {
+	return innerShuffleList(input, seed, true)
+}
+
+// UnshuffleList Un-shuffles the list by running backwards through the round count.
```
Comment (terencechain) [inline beacon-chain/utils/shuffle.go]:
```suggestion
// UnshuffleList un-shuffles the list by running backwards through the round count.
```

File: beacon-chain/utils/shuffle.go [hunk 10]
```diff
@@ -161,3 +142,140 @@ func PermutedIndex(index uint64, listSize uint64, seed [32]byte) (uint64, error)
...
+	rounds := uint8(params.BeaconConfig().ShuffleRoundCount)
+	if rounds == 0 {
+		return input, nil
+	}
+
+	listSize := uint64(len(input))
+	buf := make([]byte, totalSize, totalSize)
+	r := uint8(0)
+	if !dir {
+		// Start at last round.
```
Comment (terencechain) [inline beacon-chain/utils/shuffle.go]: Fix this similar to the comments earlier

File: beacon-chain/utils/shuffle.go [hunk 11]
```diff
@@ -161,3 +142,140 @@ func PermutedIndex(index uint64, listSize uint64, seed [32]byte) (uint64, error)
...
+			}
+			bitV := (byteV >> (j & 0x7)) & 0x1
+
+			if bitV == 1 {
+				// swap the pair items
+				input[i], input[j] = input[j], input[i]
+			}
+			//--------------------------------------------
+		}
+		// go forwards?
```
Comment (terencechain) [inline beacon-chain/utils/shuffle.go]: what does this comment mean? do we need the `?`

File: beacon-chain/utils/shuffle.go (hunk 7 above)
Comment (prestonvanloon) [reply beacon-chain/utils/shuffle.go]: These should be full sentences as Terence suggested. Please limit comment line length to 80 characters (break into a new line if comment is longer than 80 char). The latter isn't a hard rule, but helpful suggestion :)

File: beacon-chain/utils/shuffle.go [hunk 12]
```diff
@@ -145,6 +113,19 @@ func PermutedIndex(index uint64, listSize uint64, seed [32]byte) (uint64, error)
 		if bitV == 1 {
 			index = flip
 		}
+		if dir {
+			// -> shuffle
```
Comment (prestonvanloon) [inline beacon-chain/utils/shuffle.go]: What is this? This is confusing

File: beacon-chain/utils/shuffle.go [hunk 13]
```diff
@@ -145,6 +113,19 @@ func PermutedIndex(index uint64, listSize uint64, seed [32]byte) (uint64, error)
...
+			round++
+			if round == rounds {
+				break
+			}
+		} else {
+			if round == 0 {
+				break
+			}
+			// -> un-shuffle
```
Comment (prestonvanloon) [inline beacon-chain/utils/shuffle.go]: What is this? It doesn't help for readability

File: beacon-chain/utils/shuffle.go [hunk 14]
```diff
@@ -145,6 +113,19 @@ func PermutedIndex(index uint64, listSize uint64, seed [32]byte) (uint64, error)
 		if bitV == 1 {
 			index = flip
 		}
+		if dir {
```
Comment (prestonvanloon) [inline beacon-chain/utils/shuffle.go]: How about naming this `shuffle`? Then you don't need to explain that the truth case is shuffle and the false case is unshuffle in a comment.

File: beacon-chain/utils/shuffle.go [hunk 15]
```diff
@@ -161,3 +142,140 @@ func PermutedIndex(index uint64, listSize uint64, seed [32]byte) (uint64, error)
...
+
+// UnshuffleList Un-shuffles the list by running backwards through the round count.
+func UnshuffleList(input []uint64, seed [32]byte) ([]uint64, error) {
+	return innerShuffleList(input, seed, false)
+}
+
+// shuffles or unshuffles, depending on the `dir` true for shuffling, false for unshuffling.
+func innerShuffleList(input []uint64, seed [32]byte, dir bool) ([]uint64, error) {
+	if len(input) <= 1 {
+		// nothing to (un)shuffle
```
Comment (prestonvanloon) [inline beacon-chain/utils/shuffle.go]: Remove

File: beacon-chain/utils/shuffle.go [hunk 16]
```diff
@@ -161,3 +142,140 @@ func PermutedIndex(index uint64, listSize uint64, seed [32]byte) (uint64, error)
...
+	r := uint8(0)
+	if !dir {
+		// Start at last round.
+		// Iterating through the rounds in reverse, un-swaps everything, effectively un-shuffling the list.
+		r = rounds - 1
+	}
+	// Seed is always the first 32 bytes of the hash input, we never have to change this part of the buffer.
+	copy(buf[:seedSize], seed[:])
+	for {
+		// spec: pivot = bytes_to_int(hash(seed + int_to_bytes1(round))[0:8]) % list_size
```
Comment (prestonvanloon) [inline beacon-chain/utils/shuffle.go]: Please move all of this inline commentary to the godoc comment. It is too noisy and doesn't helpful readability.

File: beacon-chain/utils/shuffle.go [hunk 17]
```diff
@@ -161,3 +142,140 @@ func PermutedIndex(index uint64, listSize uint64, seed [32]byte) (uint64, error)
...
+			if j&0x7 == 0x7 {
+				byteV = source[(j&0xff)>>3]
+			}
+			bitV := (byteV >> (j & 0x7)) & 0x1
+
+			if bitV == 1 {
+				// swap the pair items
+				input[i], input[j] = input[j], input[i]
+			}
+			//--------------------------------------------
```
Comment (prestonvanloon) [inline beacon-chain/utils/shuffle.go]: Remove this useless line

File: beacon-chain/utils/shuffle.go [hunk 18]
```diff
@@ -161,3 +142,140 @@ func PermutedIndex(index uint64, listSize uint64, seed [32]byte) (uint64, error)
...
+		mirror = (pivot + listSize + 1) >> 1
+		end := listSize - 1
+		// Again, seed and round input is in place, just update the position.
+		// We start at the end, and work back to the mirror point.
+		// This makes us process each pear exactly once (instead of unnecessarily twice, like in the spec)
+		binary.LittleEndian.PutUint32(buf[pivotViewSize:], uint32(end>>8))
+		source = hashutil.HashSha256(buf)
+		byteV = source[(end&0xff)>>3]
+		for i, j := pivot+1, end; i < mirror; i, j = i+1, j-1 {
+			// Exact same thing (copy of above loop body)
```
Comment (prestonvanloon) [inline beacon-chain/utils/shuffle.go]: Copy? This is screaming for abstraction! 
DRY = Don't repeat yourself

Please refactor this code so it is abstract enough to have reusable code.

File: beacon-chain/utils/shuffle.go [hunk 19]
```diff
@@ -161,3 +142,140 @@ func PermutedIndex(index uint64, listSize uint64, seed [32]byte) (uint64, error)
...
+			//--------------------------------------------
+			// The pair is i,j. With j being the bigger of the two, hence the "position" identifier of the pair.
+			// Every 256th bit (aligned to j).
```
Comment (prestonvanloon) [inline beacon-chain/utils/shuffle.go]: This isn't a sentence and doesn't make sense

File: beacon-chain/utils/shuffle.go (hunk 18 above)
Comment (shayzluf) [reply beacon-chain/utils/shuffle.go]: thanks, fixed it

File: beacon-chain/utils/shuffle.go (hunk 14 above)
Comment (shayzluf) [reply beacon-chain/utils/shuffle.go]: good suggestion , changed

File: beacon-chain/utils/shuffle.go (hunk 16 above)
Comment (shayzluf) [reply beacon-chain/utils/shuffle.go]: moved

File: beacon-chain/utils/shuffle.go (hunk 17 above)
Comment (shayzluf) [reply beacon-chain/utils/shuffle.go]: removed

File: beacon-chain/utils/shuffle.go (hunk 19 above)
Comment (shayzluf) [reply beacon-chain/utils/shuffle.go]: changed

File: beacon-chain/utils/shuffle.go (hunk 10 above)
Comment (shayzluf) [reply beacon-chain/utils/shuffle.go]: changed

File: beacon-chain/chaintest/backend/simulated_backend.go [hunk 20]
```diff
@@ -176,17 +176,29 @@ func (sb *SimulatedBackend) RunForkChoiceTest(testCase *ForkChoiceTestCase) erro
...
-		indexList[i] = uint64(v.OriginalIndex)
+	seed := common.HexToHash(testCase.Seed)
+	testIndices := make([]uint64, testCase.Count, testCase.Count)
+	for i := uint64(0); i < testCase.Count; i++ {
+		testIndices[i] = i
 	}
-	output, err := utils.ShuffleIndices(seed, indexList)
+	shuffledList := make([]uint64, testCase.Count)
+	for i := uint64(0); i < testCase.Count; i++ {
+		si, err := utils.UnShuffledIndex(i, testCase.Count, seed)
```
Comment (nisdas) [inline beacon-chain/chaintest/backend/simulated_backend.go]: why is this unshuffled here ?
```suggestion
		si, err := utils.ShuffledIndex(i, testCase.Count, seed)
		if err != nil {
			log.Error(err)
			return err
		}
		shuffledList[i] = si
```

File: beacon-chain/utils/shuffle.go [hunk 21]
```diff
@@ -2,71 +2,24 @@
...
+	"encoding/binary"
 	"fmt"
 
-	"github.com/ethereum/go-ethereum/common"
 	"github.com/prysmaticlabs/prysm/shared/bytesutil"
 	"github.com/prysmaticlabs/prysm/shared/hashutil"
 	"github.com/prysmaticlabs/prysm/shared/params"
 )
 
+// HashFn sets the current hash function
```
Comment (terencechain) [inline beacon-chain/utils/shuffle.go]:
```suggestion
// HashFn sets the current hash function.
```

File: beacon-chain/chaintest/backend/simulated_backend.go (hunk 20 above)
Comment (shayzluf) [reply beacon-chain/chaintest/backend/simulated_backend.go]: Screen Shot 2019-05-26 at 10 22 56 AM

File: beacon-chain/utils/shuffle.go [hunk 22]
```diff
@@ -161,3 +140,99 @@ func PermutedIndex(index uint64, listSize uint64, seed [32]byte) (uint64, error)
...
+	}
+	listSize := uint64(len(input))
+	buf := make([]byte, totalSize, totalSize)
+	r := uint8(0)
+	if !shuffle {
+		r = rounds - 1
+	}
+	copy(buf[:seedSize], seed[:])
+	for {
+
```
Comment (terencechain) [inline beacon-chain/utils/shuffle.go]:
```suggestion
```

File: beacon-chain/utils/shuffle.go (hunk 21 above)
Comment (shayzluf) [reply beacon-chain/utils/shuffle.go]: its not used, i removed it

Comment (terencechain) [reply beacon-chain/utils/shuffle.go]: Huh. I still see it. We should add a period at the end

File: beacon-chain/utils/shuffle.go (hunk 7 above)
Comment (shayzluf) [reply beacon-chain/utils/shuffle.go]: thanks, i hope my latest version is better

File: beacon-chain/utils/shuffle.go (hunk 6 above)
Comment (shayzluf) [reply beacon-chain/utils/shuffle.go]: added comment for better readability

Comment (shayzluf) [reply beacon-chain/utils/shuffle.go]: thanks

File: beacon-chain/utils/shuffle.go [hunk 23]
```diff
@@ -161,3 +137,98 @@ func PermutedIndex(index uint64, listSize uint64, seed [32]byte) (uint64, error)
...
+			if r == 0 {
+				break
+			}
+			r--
+		}
+	}
+	return input, nil
+}
+
+func swapOrNot(buf []byte, byteV byte, i uint64, input []uint64, j uint64, source [32]byte) (byte, [32]byte) {
```
Comment (prestonvanloon) [inline beacon-chain/utils/shuffle.go]: Can you add a brief comment?

**Pull request merged**